Added Flyway, Eureka, configured application.properties and created db

Mishgan55 opened the pull request.

Mishgan55 committed:
Added Flyway, Eureka, configured application.properties and created db

MShablinskaya commented on src/main/resources/db/migration/V1__Init_DB.sql:
```diff
@@ -0,0 +1,43 @@
+create table branches(
```
Вроде договорились переименовать на что-то более говорящее?

MShablinskaya commented on src/main/resources/db/migration/V1__Init_DB.sql:
```diff
@@ -0,0 +1,43 @@
+                     model varchar,
+                     year_of_production integer,
+                     plate_number varchar,
+                     car_condition boolean
```
о чем говорит это поле? состояние машины сломана она или нет? поля должны быть говорящими. Если ты хочешь сетать тут доступность - так и пиши.

MShablinskaya replied:
значение по умолчанию не установлено

MShablinskaya commented on src/main/resources/db/migration/V1__Init_DB.sql:
```diff
@@ -0,0 +1,43 @@
+
+);
+
+create table customers
```
Ты будешь отдельно создавать таблицы для админов и менеджеров?

MShablinskaya commented on src/main/resources/db/migration/V1__Init_DB.sql:
```diff
@@ -0,0 +1,43 @@
+                             customer_id integer references customers(id),
+                             amount integer,
+                             creation_date timestamp
+);
```
Общее замечание по БД: не продуманы обязательные поля, уникальные поля. Подумать и переделать. Значения по умолчанию там где они необходимы.

MShablinskaya replied:
Замечание по архитектуре проекта: все пакеты именуются только в нижнем регистре. Сущности находятся в пакете entity, не модель. Для тестовых классов есть отдельная директория test? с основным проектом не мешаем. Все пакеты именуются в единственном числе. исправь.

Mishgan55 committed:
Created model, DTO, repository, service for entity Car. Fixed  db names and added constrains for fields

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/DTO/CarDTO.java:
```diff
@@ -0,0 +1,12 @@
+import lombok.Data;
+
+@Data
+public class CarDTO {
```
придерживаемся camelCase

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/controller/CarController.java:
```diff
@@ -0,0 +1,65 @@
+
+    @GetMapping
+    public List<CarDTO> index(){
+        return carService.findAll().stream().map(this::convertToCar).collect(Collectors.toList());
```
Всю бизнес логику мы выносим в сервисы. В контроллерах должна быть будквально одна строка, findAll отдающий сразу лист

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/controller/CarController.java:
```diff
@@ -0,0 +1,65 @@
+    @PostMapping("/add")
+    public ResponseEntity<HttpStatus> createCar(@RequestBody CarDTO carDTO){
+        Car car = convertToCarDTO(carDTO);
+        carService.create(car);
```
сервис должен принимать дто как аргумент, вся новертация происходит внутри сервиса

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/controller/CarController.java:
```diff
@@ -0,0 +1,65 @@
+    @DeleteMapping("/{id}")
+    public ResponseEntity<HttpStatus> delete(@PathVariable("id") Integer id){
+        carService.delete(id);
+        return ResponseEntity.ok(HttpStatus.OK);
```
update/delete - void

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/models/Car.java:
```diff
@@ -0,0 +1,53 @@
+    Boolean isAvailable;
+
+    @Override
+    public boolean equals(Object o) {
```
equals&hashCode руками не переопределяем.

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/services/CarService.java:
```diff
@@ -0,0 +1,40 @@
+
+@Service
+@Transactional(readOnly = true)
+public class CarService {
```
где CarService interface. Сохраняем флоу разработки - у нас есть интерфейс и его имплементация.

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/DTO/CarDTO.java:
```diff
@@ -0,0 +1,12 @@
+    String model;
+    Integer yearOfProduction;
+    String plateNumber;
+    Boolean isAvailable;
```
принцип инкапсуляции сохраняем в дто тоже, модификаторы доступа обязательны!

Mishgan55 committed:
Changed name of packages, some changes in class names and refactored CarController and CarService.
Created CarServiceImpl.
Created update method for Car.

Mishgan55 committed:
1.Deleted ModelMapper and added MapStruct
2.Added uniq validation for plate number
3.Added Exception handlers and handled all errors
4. Added validation on the fields
5. Added CarServiceImpl

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/service/CarService.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.Optional;
+
+@Service
+public interface CarService<T> {
```
Сам клас можешь не дженерализировать, а вот в методах лучше указать дтошку, вместо абстрактного Т

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/mapper/CarMapper.java:
```diff
@@ -0,0 +1,13 @@
+
+@Mapper(componentModel = "spring")
+public interface CarMapper {
+    CarMapper INSTANCE = Mappers.getMapper( CarMapper.class );
```
instance не используем. рбаотай с маппером через создание бинов, как и с любым другим компонентом.

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/service/impl/CarServiceImpl.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+
+    private void updatingCarFields(Car car, Integer id, CarDto carDto) {
```
Для этих целей мы используем маппер.

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/service/CarService.java:
```diff
@@ -0,0 +1,15 @@
+
+@Service
+public interface CarService<T> {
+    Optional<T> showOne(int id);
```
тут чисто дело вкуса, но я по названию бы подумала, что их найдется много по ID. можно обозвать что-то типа findCarDtoById.

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/service/CarService.java:
```diff
@@ -0,0 +1,15 @@
+    Optional<T> showOne(int id);
+    List<T> findAll();
+    void create(T car);
+    void delete(int id);
```
Почему входной параметр примитив? А где-то обертка. Лучше использовать обертку, легче проверить на null.

Mishgan55 committed:
Refactored Mapper and all method for mapper
Refactored CarService

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/controller/CarController.java:
```diff
@@ -0,0 +1,59 @@
+@RequestMapping("/car")
+public class CarController {
+
+    private final CarServiceImpl carService;
```
объявляем интерфейс, а не реализацию.

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/service/impl/CarServiceImpl.java:
```diff
@@ -0,0 +1,84 @@
+    @Transactional
+    public void updateCar(Integer id, CarDto carDto) {
+
+        Optional<CarDto> updatedCarDto = findCarDtoById(id);
```
неверная логика

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/service/impl/CarServiceImpl.java:
```diff
@@ -0,0 +1,84 @@
+        }
+    }
+
+    private void enrichCarForUpdate(Car car, Integer id) {
```
Это что?

MShablinskaya replied:
Это что?

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/service/impl/CarServiceImpl.java:
```diff
@@ -0,0 +1,84 @@
+        return carRepository.findById(id).map(carMapper::carToCarDto);
+    }
+
+    public List<CarDto> findAllCar() {
```
cars

Mishgan55 committed:
Refactored update method for Car

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/controller/CarController.java:
```diff
@@ -0,0 +1,60 @@
+
+    @GetMapping
+    public List<CarDto> index() {
+        return carService.findAllCar();
```
public ResponseEntity<List<CarDto>> getAllCars() {
        return ResponseEntity.ok(carService.getAllCars());
    }
методы в контроллере(да и любые методы) - это глаголы или начинаются с глагола!

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/controller/CarController.java:
```diff
@@ -0,0 +1,60 @@
+    @PostMapping("/add")
+    public ResponseEntity<HttpStatus> createCar(@RequestBody @Valid CarDto carDTO) {
+        carService.createCar(carDTO);
+        return ResponseEntity.ok(HttpStatus.CREATED);
```
добавить, обновить, удалить - void!

MShablinskaya commented on src/main/java/innowise/khorsun/carorderservice/service/CarService.java:
```diff
@@ -0,0 +1,17 @@
+    void createCar(CarDto carDto);
+    void deleteCar(Integer id);
+    void updateCar(Integer id, CarUpdateDto carUpdateDto);
+
```
если в репозитории находишь, сохраняеш  и тд, то в сервисе ты получаешь(getCarById, getAllCars и тд)

Mishgan55 committed:
Renamed methods and refactored controller

Mishgan55 committed:
Renamed methods and refactored controller Number 2

Mishgan55 committed:
Refactored method getOneCar
